outlines/index/index.py

```diff
@@ -0,0 +1,101 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, List, NewType, Optional, Protocol
+
+import interegular
+import torch
+
+from outlines.index.fsm import create_fsm_index_tokenizer, make_deterministic_fsm
+
+if TYPE_CHECKING:
+    from outlines.models.tokenizer import Tokenizer
+
+FSMState = NewType("FSMState", int)
+
+
+@dataclass(frozen=True)
+class GenerateInstruction:
+    tokens_to_mask: List[int]
+
+
+class FSM(Protocol):
+    def next_instruction(self, state: FSMState) -> GenerateInstruction:
+        ...
+
+    def next_state(self, state: FSMState, token_id: torch.Tensor) -> FSMState:
+        ...
+
+    def is_final_state(self, state: FSMState) -> bool:
+        ...
+
+
+class StopAtTokenFSM:
```

Comment:
We can implement a map similar to the one used in `Regex`, if there is a simple way to represent "all tokens"